### ISSUE TITLE
Best practices for multilingual values

### ISSUE THREAD

**Issue by ajs6f** [w3c/json-ld-syntax#105]
As evidenced by [reports](https://github.com/w3c/json-ld-syntax/issues/91), there is some confusion about how to use multilingual data values alongside language maps. @pchampin [noted](https://github.com/w3c/json-ld-syntax/issues/91#issuecomment-445313785) that using an alias is a good way to work through this, and @BigBlueHat noted (link to minutes forthcoming) that this is the approach taken by Web Annotations. We should offer some examples of this practice, probably in the context of the (long-promised) Primer.

**Comment by gkellogg** [comment]
Riffing off of @pchampin's example in https://github.com/w3c/json-ld-syntax/issues/91#issuecomment-445313785, we might use data indexing to aid access:

```json
{
  "@context": {
    "occupation": { "@id": "ex:occupation", "@type": "rdf:HTML", "@container": "@data" },
    "description": "ex:description"
  },
  "name": "Yagyū Muneyoshi",
  "occupation": {
    "ja": "<span lang="en">Ninja in japanese: <span lang="jp">忍者</span>",
    "en": "<span lang="en">Ninja in english: <span lang="en">Ninja</span>",
    "cs": "<span lang="en">Ninja in czech: <span lang="cs"> Nindža </span>"
  }
}
```

This allows data indexing and consistent use of HTML values.

**Comment by iherman** [comment]
But... what would the generated RDF look like? One cannot add a language tag to a typed literal:-(

**Comment by gkellogg** [comment]
Its not a language tag, it’s a data index which has no RDF representation.

**Comment by iherman** [comment]
Oops... well, this is one of those surprise effect that @ajs6f was talking about yesterday: I missed the `"@container": "@data"` and thought it was language. Ie, if I am an author not looking into the details of the context, I can be a bit misled.

Yes, it is legal; I do not think it is good practice.

**Comment by BigBlueHat** [comment]
seeAlso https://iiif.io/api/presentation/3.0/#44-html-markup-in-property-values

**Comment by iherman** [comment]
This issue was discussed in [a meeting](https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2019/2019-04-05-json-ld#section4-2).

- `RESOLVED:  highlight the need for work is ongoing, but it should present what can be done today via language/data maps and/or using HTML (or other) micro-syntax for expressing multiple language`
<details><summary><i class="differentiate">View the transcript</i></summary>
Multilingual Values<br/>
Benjamin Young: <a href="https://github.com/w3c/json-ld-syntax/issues/105">https://github.com/w3c/json-ld-syntax/issues/105</a><br/>
Benjamin Young: Another easy one ;)<br/>
… this one is about how JSON-LD currently works, and our past decisions to use HTML for multi lingual values (strings with multiple languages)<br/>
… so use straight up HTML, which is not ideal<br/>
… Looking at text level semantics HTML, but that’s for the future.<br/>
… so what do we need to propose in the primer to close the issue?<br/>
… related - there’s no way to do multi-language language maps<br/>
Rob Sanderson: it seems we should split this into a primer issue<br/>
… eg how do you use language tags<br/>
… and what do you do with multiple languages<br/>
… and then have a syntax issue around gkellogg’s issue for the normative specs<br/>
Benjamin Young: …about, is it an error to have English and Japanese in a string that is stated to be only one of those<br/>
Ivan Herman: What was put there by gregg sounds like a solution, but a bit misleading. The use of language tags gives the wrong impression — should be just indexes<br/>
Ivan Herman: Language tags are defined by ISO<br/>
Rob Sanderson: “&lt;span lang="en">Ninja in japanese: <span lang="jp">忍者</span>“@ja<br/>
Rob Sanderson: I agree ivan. to your question, the RDF would look like that:<br/>
Rob Sanderson: "<span lang='en'>Ninja in japanese: <span lang='jp'>忍者</span>"@ja^^rdf:langString<br/>
Rob Sanderson: this has been my issue for 5+ years<br/>
… language tags must be langString<br/>
Ivan Herman: an RDF issue that is not ours to solve<br/>
… Lots of nice discussions in dbooth’s repo, but it should happen in RDF not here<br/>
… same as missing base direction<br/>
… we can only set a single language. And this is the same as base direction, shouldn’t touch it<br/>
Rob Sanderson: +1 to ivan<br/>
Benjamin Young: RDF is woefully broken in this way, but Gregg’s proposal of HTML + language map would be desirable by JSON developers<br/>
Rob Sanderson: <a href="https://iiif.io/api/presentation/3.0/#44-html-markup-in-property-values">https://iiif.io/api/presentation/3.0/#44-html-markup-in-property-values</a><br/>
Benjamin Young: If built to contain HTML, they’re not going to take it into RDF, so a little misuse has advantages<br/>
Ivan Herman: q=<br/>
Benjamin Young: our audience is interested in JSON, with a side plate of a graph<br/>
Rob Sanderson: I put this link in earlier <a href="https://iiif.io/api/presentation/3.0/#44-html-markup-in-property-values">https://iiif.io/api/presentation/3.0/#44-html-markup-in-property-values</a><br/>
… it uses exactly what gkellogg describes<br/>
… it is common and exactly what people want to be able to do<br/>
Ivan Herman: The funny thing is what you wrote is legal but ugly RDF – a microsyntax for a string, which is outside of RDF or JSON-LD<br/>
… it happens to be a subsyntax of HTML<br/>
… don’t need anything in the syntax document to do this, its a private agreement between parties<br/>
Rob Sanderson: +1 to Ivan<br/>
Ivan Herman: this is probably the only thing we can do<br/>
… so no issue in the syntax document<br/>
… it’s an ugly but best practice given the current technologies<br/>
Pierre-Antoine Champin: Going to propose a crazy idea, in the line of what Ivan said. We don’t need to change RDF, we could define a custom datatype. langString is syntactic sugar for a standard datatype for a more ugly microsyntax of the language inside the value<br/>
… we could define a more complex but similar datatype. That’s the crazy idea :) We could instrument it in RDF, with another container type, so that what gregg proposed would generate the appropriate structure<br/>
… but it’s quite some work<br/>
Ivan Herman: technically … yes … and now I put on the W3C hat, it’s outside of our charter. This would be a RDF datatype.<br/>
Pierre-Antoine Champin: What about JSON data type?<br/>
Ivan Herman: JSON is closer to our charter. But language isn’t.<br/>
… it would be a lot of work … the flood gates would be open. Ruby, direction, etc.<br/>
Benjamin Young: <a href="https://w3c.github.io/string-meta/">https://w3c.github.io/string-meta/</a><br/>
Benjamin Young: worth pausing on the JSON data type. I hear the concerns … is there a way around them? This string-meta document from i18n suggests JSON-LD as a solution for multi-language use<br/>
… feel that there’s an opportunity here<br/>
… And if we miss it, there’ll be a lot of terrible looking JSON-LD<br/>
… I see that it evokes process specters, but it comes up a lot<br/>
… The genie won’t go back into the bottle. So any hope of this?<br/>
Ivan Herman: Don’t remember the issue, but got into a long discussion with the editors. The examples are mostly wrong.<br/>
Benjamin Young: <a href="https://github.com/w3c/string-meta/issues/27">https://github.com/w3c/string-meta/issues/27</a><br/>
Benjamin Young: also <a href="https://github.com/w3c/string-meta/issues/13">https://github.com/w3c/string-meta/issues/13</a><br/>
Ivan Herman: I understand the problem. Would love for the problem to be solved, but outside our influence<br/>
Benjamin Young: oh…and <a href="https://github.com/w3c/string-meta/issues/23">https://github.com/w3c/string-meta/issues/23</a><br/>
Ivan Herman: I don’t see any other proper way, other than having it done at the RDF level.<br/>
Benjamin Young: …and another <a href="https://github.com/w3c/string-meta/issues/11">https://github.com/w3c/string-meta/issues/11</a><br/>
Rob Sanderson: The bigger risk is to build on shifting sands and have RDF come up with a different syntax that’s incompatible with whatever we come up with<br/>
… should instead use it as a way to highlight the need, and potentially a micro-chartered group to solve it for RDF<br/>
Benjamin Young: Not ready to recharter, or make a new datatype. Rob proposes to kick it to another group and then an update to JSON-LD. Not a solution, but don’t want to lose the actions<br/>
… to close the issue we should state what can be done<br/>
… but need to be clear as to what /should/ be done that’s not confusing<br/>
Jeff Mixter: +1 to that<br/>
Proposed resolution: highlight the need for work is ongoing, but it should present what can be done today via language/data maps and/or using HTML (or other) micro-syntax for expressing multiple language (Benjamin Young)<br/>
Rob Sanderson: +1<br/>
Benjamin Young: +1<br/>
Jeff Mixter: +1<br/>
Ivan Herman: +1<br/>
Tim Cole: +1<br/>
Pierre-Antoine Champin: +1<br/>
Simon Steyskal: +1<br/>
Adam Soroka: +1<br/>
Resolution #5: highlight the need for work is ongoing, but it should present what can be done today via language/data maps and/or using HTML (or other) micro-syntax for expressing multiple language<br/>
Ivan Herman: procedural question - if we close the issue, then I think we will lose it for the bp doc. For the time being we don’t have an editor for the document. So don’t want it lost.<br/>
… should be raised in the BP repo<br/>
Rob Sanderson: +1<br/>
Benjamin Young: +1<br/>
Ivan Herman: should go through the issues to make sure we don’t lose them<br/>
Benjamin Young: Agreed – open editorial issues on BP?<br/>
… keep these initial discussion in the syntax doc, to not have the comments scattered<br/>
Ivan Herman: Wouldn’t close this one<br/>
Simon Steyskal: <a href="https://github.com/w3c/json-ld-bp/issues">https://github.com/w3c/json-ld-bp/issues</a><br/>
Benjamin Young: not until there’s another issue to write it up<br/>
Ivan Herman: editor will write it up as they see best<br/>
Benjamin Young: And it’s the top of the hour<br/>
… thanks for all the input<br/>
</details>
